zxlive/proof_actions.py

```diff
@@ -112,17 +115,90 @@ def rewriter() -> None:
             return rewriter
         for action in self.actions:
             if action.button is not None: continue
-            btn = QPushButton(action.name,parent)
+            btn = QPushButton(action.name, parent)
             btn.setMaximumWidth(150)
             btn.setStatusTip(action.tooltip)
             btn.setEnabled(False)
-            btn.clicked.connect(create_rewrite(action,parent))
+            btn.clicked.connect(create_rewrite(action, parent))
             self.btn_group.addButton(btn)
             action.button = btn
 
     def update_active(self, g: GraphT, verts: List[VT], edges: List[ET]) -> None:
         for action in self.actions:
-            action.update_active(g,verts,edges)
+            action.update_active(g, verts, edges)
+
+
+def to_networkx(graph: Graph) -> nx.Graph:
+    G = nx.Graph()
+    v_data = {v: {"type": graph.type(v),
+                  "phase": graph.phase(v),
+                  "boundary_index": graph.vdata(v, "boundary_index", default=-1),}
+              for v in graph.vertices()}
+    G.add_nodes_from([(v, v_data[v]) for v in graph.vertices()])
+    G.add_edges_from([(*v, {"type": graph.edge_type(v)}) for v in  graph.edges()])
+    return G
+
+def create_subgraph(graph: Graph, verts: List[VT]) -> nx.Graph:
+    graph_nx = to_networkx(graph)
+    subgraph_nx = nx.Graph(graph_nx.subgraph(verts))
+    boundary_mapping = {}
+    i = 0
+    for v in verts:
+        for vn in graph.neighbors(v):
+            if vn not in verts:
+                boundary_node = 'b' + str(i)
+                boundary_mapping[boundary_node] = vn
+                subgraph_nx.add_node(boundary_node, type=VertexType.BOUNDARY)
+                subgraph_nx.add_edge(v, boundary_node, type=EdgeType.SIMPLE)
+                i += 1
+    return subgraph_nx, boundary_mapping
+
+def custom_matcher(graph: Graph, in_selection: Callable[[VT], bool], lhs_graph: nx.Graph) -> List[VT]:
+    verts = [v for v in graph.vertices() if in_selection(v)]
+    subgraph_nx, _ = create_subgraph(graph, verts)
+    graph_matcher = GraphMatcher(lhs_graph, subgraph_nx,\
+        node_match=categorical_node_match(['type', 'phase'], default=[1, 0]))
+    if graph_matcher.is_isomorphic():
+        return verts
+    return []
+
+def custom_rule(graph: Graph, vertices: List[VT], lhs_graph: nx.Graph, rhs_graph: nx.Graph) -> pyzx.rules.RewriteOutputType[ET,VT]:
```

Comment:
So if I understand the logic of this function, you remove all the vertices in the original graph, and then you replace them with new ones (except for the boundaries). While this does give you correct semantics, I think in terms of positioning information we will want to use something that tries to match the lhs and the rhs as match as possible, and only adds/removes those vertices that really have to be removed.
tl;dr: The matching graphs we are using have position information, and this right now is not being used.

Reply:
This might not be the thing to do for the first draft, but some kind of positioning must be done, because you dont want them all to appear just on top of each other.

Reply:
I was wondering about this. Where is this position information? I don't see anything obvious in pyzx Graph data structure.

Reply:
you can call g.get_pos(v) I believe, or otherwise g.get/set_qubit(v), which corresponds to the y coordinate and g.get/set_row(v) which corresponds to the x coordinate.

Reply:
That did not work for me. I got the attribute error: `'GraphS' object has no attribute 'get_pos'`. Same for `get_qubit` and `get_row`.